### DISCUSSION TITLE
Bug: useInterpret: new subscription at each render

### DISCUSSION THREAD

**Discussion by mathieutu** [statelyai/xstate#3536] [Q&A] [answered]
### Description

Hi, when passing a listener to `useInterpreter` like shown in [the doc](https://xstate.js.org/docs/packages/xstate-react/#useinterpret-machine-options-observer), the function is recreated at each render.

Nothing problematic as such, but the listener is passed to the dependencies of the useEffect which will open the subscription:
https://github.com/statelyai/xstate/blob/7d5be188f85bca4d25d79888d4e14bdd6811952e/packages/xstate-react/src/useInterpret.ts#L125-L135

So each time the component re-render, the subscription is re openened, and the function is recalled.

We could remove it from dependencies, or use I think (not sure) the new useSyncExternalStore hook.

We also can update the doc to recommend the usage of `useCallback`.

Thanks.

### Expected result

Stable listener, called only when the state changes.

### Actual result

Listener called at each render.

### Reproduction

https://codesandbox.io/s/sad-antonelli-twx2mv?file=/src/index.js

### Additional context

Thanks! 👋

**Comment by davidkpiano** [comment, marked as answer]
It's probably causing a new subscription because the subscription function you're passing in is not a stable reference.

Try this:
```js
  const logValue = React.useCallback((state) => {
    console.log(state.value);
  }, []);

  const service = useInterpret(toggleMachine, {}, logValue);
```

**Reply by mathieutu** [reply]
Hey @davidkpiano, thanks for answering, 

>not a stable reference

Yeah, it's indeed what I said in the issue 😅 . 

The code is taken from the [xstate/react doc](https://xstate.js.org/docs/packages/xstate-react/#useinterpret-machine-options-observer), if `useCallback` is needed on the user side, we should update the doc then?

**Reply by davidkpiano** [reply]
Sorry, read the issue a little too quickly!

Yes, the documentation should be updated.

**Reply by mathieutu** [reply]
Ok, I'll add a PR in my todo list then 🙂
